Jetpack Pro Dashboard: implement monitor notification settings modal

yashwin opened the pull request:
#### Proposed Changes

This PR implements monitor notification settings modal in the Jetpack Pro Dashboard.

#### Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

> **_NOTE:_** Changing the monitor status will not update the status immediately; we will have to debug this - 1203448324265423-as-1203576555374768. Validations & API integration to save the notification settings will be done in a separate PR.

**Instructions**

1. Run `git checkout add/implement-monitor-notification-settings-modal` and `yarn start-jetpack-cloud`
2. Open http://jetpack.cloud.localhost:3000/, and you'll be redirected to the /dashboard.
3. Click the clock icon(which also shows the current notification schedule) or Enable the monitor for any site by clicking the toggle on the monitor column if you don't have monitor enabled on any of your sites.
4. Please note the monitor will not be enabled immediately; you might have to wait for some time until it is enabled. 
5. Once clicked on the clock icon, verify that the notification settings popup is shown as below.

<img width="492" alt="Screenshot 2022-12-21 at 12 47 46 PM" src="https://user-images.githubusercontent.com/10586875/208847557-cd02e6d4-8494-4f24-9405-006f23c3a844.png">

<img width="560" alt="Screenshot 2022-12-21 at 12 48 43 PM" src="https://user-images.githubusercontent.com/10586875/208847610-1bdfa73c-ccb9-4c96-ad79-743df8c3adb4.png">

<img width="520" alt="Screenshot 2022-12-21 at 12 48 58 PM" src="https://user-images.githubusercontent.com/10586875/208847630-b2e4b39b-7f41-4591-8a2a-7cf119b2f5e2.png">

Mobile view

<img width="539" alt="Screenshot 2022-12-21 at 12 50 56 PM" src="https://user-images.githubusercontent.com/10586875/208848249-a22c1a8f-566d-4579-b971-a81873089e52.png">

Tablet view

<img width="494" alt="Screenshot 2022-12-21 at 12 59 43 PM" src="https://user-images.githubusercontent.com/10586875/208848310-9a036fbb-3231-4c3c-8ae9-eb6d4c8dc404.png">

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to 1203448324265423-as-1203507036638783

yashwin committed:
Implement monitor notification settings modal

yashwin committed:
Fix table responsiveness for monitor

yashwin committed:
Update aria-label for notification schedule

yashwin commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notifcation-settings/index.tsx:
```diff
@@ -0,0 +1,192 @@
+	const [ selectedDuration, setSelectedDuration ] = useState< Duration | undefined >( undefined );
+	const [ addedEmailAddresses, setAddedEmailAddresses ] = useState< string[] | [] >( [] );
+
+	const mobileAppLink = 'https://jetpack.com/blog/jetpack-mobile-app/';
```
@madebynoam - Is this link ok?

madebynoam replied:
Ooh, good question, we have a slightly newer page here that we could use:
https://jetpack.com/mobile/

yashwin replied:
Sure, thanks :)

yashwin replied:
Updated here - https://github.com/Automattic/wp-calypso/pull/71373/commits/5ba3e76f146eeb6a6fc20651f313953d74ba45e0

DustinHartzler commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notifcation-settings/index.tsx:
```diff
@@ -0,0 +1,192 @@
+				<div className="notification-settings__small-screen">{ addEmailsContent }</div>
+			</div>
+			<div className="notification-settings__footer">
+				<Button
+					onClick={ onClose }
+					aria-label={ translate( 'Cancel and close notification settings popup' ) }
+				>
+					{ translate( 'Cancel' ) }
+				</Button>
```
Along with the Cancel button and `X` for closing the modal, can we also dismiss the popup with the Escape key or clicking outside the modal?

yashwin replied:
I purposely added `shouldCloseOnClickOutside={ false }` since this modal is something that has user-filled data, which might get lost if the user clicks outside by mistake. I don't have a very strong opinion on this. 

@madebynoam - any thoughts here?

madebynoam replied:
That is another great question. At first, I also thought it would be better to disable Escape, but it seems to be more accessible to allow it.  So let's go ahead and enable it.
https://www.w3.org/WAI/ARIA/apg/example-index/dialog-modal/dialog#accessibilityfeatures
https://clearleft.com/thinking/accessibility-in-modal-dialogs

As for clicking outside, I think it's also ok to allow it (since it's similar to escape).

yashwin replied:
Makes sense. Thanks, @madebynoam :)

yashwin replied:
Looks like it was already getting closed on clicking ESC, I just changed it to close on clicking outside - https://github.com/Automattic/wp-calypso/pull/71373/commits/5ba3e76f146eeb6a6fc20651f313953d74ba45e0

DustinHartzler commented:
Is the `Save` button in the modal supposed to work? I was able to interact with the elements in the modal, but the Save did nothing.

yashwin commented:
> Is the `Save` button in the modal supposed to work? I was able to interact with the elements in the modal, but the Save did nothing.

@DustinHartzler - We will handle the API integration for `Save` in a different PR.

yashwin committed:
Use form for notification popup

yashwin committed:
Move durations to utils

yashwin committed:
Move mobile app link to utils

yashwin committed:
Fix typo in file directory

yashwin committed:
Allow close modal on clicking outside and change Jetpack mobile link

madebynoam reviewed (approved):
Lgtm!

vitozev commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/utils.ts:
```diff
@@ -414,3 +414,32 @@ export const getProductSlugFromProductType = ( type: string ): string | undefine
+	{
+		time: 1,
+		label: translate( 'After 1 minute' ),
+	},
```
Can we remove the `1 minute` option, as it won't be available for now?

vitozev commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx:
```diff
@@ -0,0 +1,168 @@
+export default function NotificationSettings( { onClose, site, settings }: Props ) {
+	const translate = useTranslate();
+
+	const [ enableMobileNotification, setEnablMobileNotification ] = useState< boolean >( false );
```
I think `setEnablMobileNotification` has a typo. It should be `setEnableMobileNotification`.

vitozev commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx:
```diff
@@ -0,0 +1,168 @@
+		setSelectedDuration( duration );
+	}
+
+	useEffect( () => {
+		if ( settings?.monitor_deferment_time ) {
+			const foundDuration = durations.find(
+				( duration ) => duration.time === settings.monitor_deferment_time
+			);
+			foundDuration && setSelectedDuration( foundDuration );
+		}
+	}, [ settings?.monitor_deferment_time ] );
```
I like that we are using multiple `useEffect` here for the separation of concerns, as suggested in [reactjs.org.](https://reactjs.org/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns)

What does this mean in terms of performance? Is it better to have one vs. multiple ones? cc @robertf4 I'll appreciate your thoughts on this one.

yashwin replied:
I feel like it is completely fine to use multiple `useEffect`s.

vitozev commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx:
```diff
@@ -0,0 +1,168 @@
+	const addEmailsContent = enableEmailNotification && (
+		<div className="notification-settings__email-container">
+			<TokenField
+				placeholder="Enter Email IDs"
```
Can we use `translate` for this placeholder?

vitozev commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx:
```diff
@@ -0,0 +1,168 @@
+								<img
+									className="notification-settings__duration-icon"
+									src={ clockIcon }
+									alt="Durations"
```
Can we use `translate` for this `alt` attribute?

vitozev commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/toggle-activate-monitoring/index.tsx:
```diff
@@ -54,22 +62,63 @@ export default function ToggleActivateMonitoring( { site, status, settings }: Pr
+								comment:
+									'%(currentSchedule) is the current notification duration set, e.g. "1 hour" or "5 minutes"',
+							}
+						) as string
```
Out of curiosity: why are we casting the returned value of `translate` to string here?

yashwin replied:
`<Button>` component expects the aria-label to be a string, whereas the translate is not considered a string but a ReactChild.

vitozev commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/toggle-activate-monitoring/index.tsx:
```diff
@@ -54,22 +62,63 @@ export default function ToggleActivateMonitoring( { site, status, settings }: Pr
+						) as string
+					}
+				>
+					<img src={ clockIcon } alt="Notification Settings" />
```
Can we use `translate` for this `alt` attribute?

vitozev reviewed (commented):
Awesome work, @yashwin! This PR looks and works great on all device sizes! I've left some minor feedback.

I also have one UX question: how easy would it be to allow `Space` as a delimiter? I think it's a pretty common UX pattern, and if we can easily integrate it, that would be great.

yashwin committed:
Minor code changes

yashwin committed:
Tokenize on space for add email

yashwin commented:
Thanks for your review, @vitozev. I have addressed all your comments. 

> I think it's a pretty common UX pattern, and if we can easily integrate it, that would be great.

Yes, agree. I have added this support too.

vitozev reviewed (approved):
LGTM!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7688748

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7688748).
